File: agents/windows-winui/Services/ReportClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;

namespace LiveDashboardAgent.Services;

public sealed class ReportClient : IDisposable
{
    private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(10);
    private static readonly TimeSpan MaxBackoff = TimeSpan.FromSeconds(60);
    private static readonly TimeSpan PauseDuration = TimeSpan.FromMinutes(5);
    private const int PauseAfterFailures = 5;

    private readonly HttpClient _httpClient;
    private readonly Uri _endpoint;
    private int _consecutiveFailures;
    private TimeSpan _currentBackoff = TimeSpan.Zero;
    private DateTimeOffset _pauseUntil = DateTimeOffset.MinValue;

    public ReportClient(AppConfig config)
    {
        var normalized = config.Normalize();
        _endpoint = new Uri(normalized.ServerUrl.TrimEnd('/') + "/api/report", UriKind.Absolute);
        _httpClient = new HttpClient
        {
            Timeout = RequestTimeout,
        };
        _httpClient.DefaultRequestHeaders.Authorization =
            new AuthenticationHeaderValue("Bearer", normalized.Token);
    }

    public TimeSpan RetryDelay => PauseRemaining > TimeSpan.Zero ? PauseRemaining : _currentBackoff;

    public async Task<ReportResult> SendAsync(ReportPayload payload, CancellationToken cancellationToken)
    {
        if (PauseRemaining > TimeSpan.Zero)
        {
            return ReportResult.Failed("上报暂停中。");
        }

        try
        {
            using var response = await _httpClient.PostAsJsonAsync(_endpoint, payload, cancellationToken)
                .ConfigureAwait(false);
            if ((int)response.StatusCode is 200 or 201 or 409)
            {
                _consecutiveFailures = 0;
                _currentBackoff = TimeSpan.Zero;
                _pauseUntil = DateTimeOffset.MinValue;
                return ReportResult.Sent();
            }

            var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            RegisterFailure();
            return ReportResult.Failed($"HTTP {(int)response.StatusCode}: {body[..Math.Min(body.Length, 200)]}");
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception ex)
        {
            RegisterFailure();
            return ReportResult.Failed(ex.Message);
        }
    }

    public void Dispose()
    {
        _httpClient.Dispose();
    }

    private TimeSpan PauseRemaining
    {
        get
        {
            var remaining = _pauseUntil - DateTimeOffset.UtcNow;
            if (remaining <= TimeSpan.Zero)
            {
                _pauseUntil = DateTimeOffset.MinValue;
                return TimeSpan.Zero;
            }
            return remaining;
        }
    }

    private void RegisterFailure()
    {
        _consecutiveFailures++;
        _currentBackoff = _currentBackoff == TimeSpan.Zero
            ? TimeSpan.FromSeconds(5)
            : TimeSpan.FromSeconds(Math.Min(_currentBackoff.TotalSeconds * 2, MaxBackoff.TotalSeconds));

        if (_consecutiveFailures < PauseAfterFailures)
        {
            return;
        }

        _pauseUntil = DateTimeOffset.UtcNow + PauseDuration;
        _consecutiveFailures = 0;
        _currentBackoff = TimeSpan.Zero;
    }
}

public sealed record ReportResult(bool Ok, string Error)
{
    public static ReportResult Sent() => new(true, "");
    public static ReportResult Failed(string error) => new(false, error);
}

File: agents/windows-flutter/lib/src/services/report_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import '../models/app_config.dart';

class ReportClient {
  ReportClient(AppConfig config)
    : _config = config.normalize(),
      _client = HttpClient() {
    _client.connectionTimeout = const Duration(seconds: 10);
  }

  static const _requestTimeout = Duration(seconds: 10);
  static const _maxBackoff = Duration(seconds: 60);
  static const _pauseAfterFailures = 5;
  static const _pauseDuration = Duration(minutes: 5);

  final AppConfig _config;
  final HttpClient _client;
  int _consecutiveFailures = 0;
  Duration _currentBackoff = Duration.zero;
  DateTime _pauseUntil = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);

  Duration get retryDelay {
    final remaining = _pauseUntil.difference(DateTime.now().toUtc());
    if (remaining.isNegative || remaining == Duration.zero) {
      return _currentBackoff;
    }
    return remaining;
  }

  Future<ReportResult> send({
    required String appId,
    required String windowTitle,
    required Map<String, Object?> extra,
  }) async {
    if (retryDelay > _currentBackoff) {
      return const ReportResult(false, '上报暂停中。');
    }
    try {
      final endpoint = Uri.parse(
        '${_config.serverUrl.replaceFirst(RegExp(r'/+$'), '')}/api/report',
      );
      final request = await _client.postUrl(endpoint).timeout(_requestTimeout);
      request.headers.contentType = ContentType.json;
      request.headers.set(
        HttpHeaders.authorizationHeader,
        'Bearer ${_config.token}',
      );
      request.write(
        jsonEncode({
          'app_id': appId,
          'window_title': windowTitle.length > 256
              ? windowTitle.substring(0, 256)
              : windowTitle,
          'timestamp': DateTime.now().toUtc().toIso8601String(),
          'extra': extra,
        }),
      );
      final response = await request.close().timeout(_requestTimeout);
      final body = await utf8.decodeStream(response).timeout(_requestTimeout);
      if (response.statusCode == 200 ||
          response.statusCode == 201 ||
          response.statusCode == 409) {
        _consecutiveFailures = 0;
        _currentBackoff = Duration.zero;
        _pauseUntil = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
        return const ReportResult(true, '');
      }
      _registerFailure();
      return ReportResult(
        false,
        'HTTP ${response.statusCode}: ${body.substring(0, body.length > 200 ? 200 : body.length)}',
      );
    } catch (error) {
      _registerFailure();
      return ReportResult(false, error.toString());
    }
  }

  void close() {
    _client.close(force: true);
  }

  void _registerFailure() {
    _consecutiveFailures += 1;
    _currentBackoff = _currentBackoff == Duration.zero
        ? const Duration(seconds: 5)
        : Duration(
            seconds: (_currentBackoff.inSeconds * 2).clamp(
              5,
              _maxBackoff.inSeconds,
            ),
          );
    if (_consecutiveFailures >= _pauseAfterFailures) {
      _pauseUntil = DateTime.now().toUtc().add(_pauseDuration);
      _consecutiveFailures = 0;
      _currentBackoff = Duration.zero;
    }
  }
}

class ReportResult {
  const ReportResult(this.ok, this.error);

  final bool ok;
  final String error;
}
